File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    vector<vector<string>>ans;
    vector<string>board;
    set<int>colSet;
    set<int>posDia;
    set<int>negDia;
    vector<vector<string>> solveNQueens(int n) {
        board = vector<string>(n, string(n, '.'));
        backtrack(0,n);
        return ans;
    }
    void backtrack(int r, int n){
        if(r==n){
            ans.push_back(board);
            return;
        }
        for(int c=0;c<n;c++){
            if(colSet.contains(c) || posDia.contains(r+c)||negDia.contains(r-c)){
                continue;
            }
            colSet.insert(c);
            posDia.insert(r+c);
            negDia.insert(r-c);
            board[r][c]='Q';
            backtrack(r+1,n);
            colSet.erase(c);
            posDia.erase(r+c);
            negDia.erase(r-c);
            board[r][c]='.';
        }
    }
};

File: 0051-n-queens/0051-n-queens.java
class Solution {
    List<List<String>>res = new ArrayList<>();
    List<List<String>>board = new ArrayList<>();
    Set<Integer>cols = new HashSet<>();
    Set<Integer>posDiag = new HashSet<>();
    Set<Integer>negDiag = new HashSet<>();
    public List<List<String>> solveNQueens(int n) {
        for(int i =0;i<n;i++){
            List<String>temp = new ArrayList<>();
            for(int j=0;j<n;j++){
                temp.add(".");
            }
            board.add(temp);
        }
        backtrack(0,n);
        return res;
    }
    public void backtrack(int r,int n){
        if(r == n){
            // Creating a List to store the concatenated strings
        List<String> copy = new ArrayList<>();

        // Concatenating strings in each row
        for (List<String> row : board) {
            copy.add(String.join("", row));
        }
            res.add(copy);
            return;
        }
        for(int c =0;c<n;c++){
            if (cols.contains(c) || posDiag.contains(r+c) || negDiag.contains(r-c)){
                continue;
            }
            cols.add(c);
            posDiag.add(r+c);
            negDiag.add(r-c);
            board.get(r).set(c,"Q");
            backtrack(r+1,n);
            cols.remove(c);
            posDiag.remove(r+c);
            negDiag.remove(r-c);
            board.get(r).set(c,".");
        }
    }
}